samza-kv/src/main/scala/org/apache/samza/storage/kv/AccessLoggedStore.scala

```diff
@@ -0,0 +1,155 @@
+  val systemStream = new SystemStream(changelogSystemStreamPartition.getSystemStream.getSystem, streamName)
+  val partitionId: Int = changelogSystemStreamPartition.getPartition.getPartitionId
+  val serializer = new LongSerde()
+  val samplingRatio = storageConfig.getAccessLogSamplingRatio(storeName)
+  val rng = scala.util.Random
+
+  def get(key: K): V = {
+    val isRange = false
+    measureLatencyAndWriteToStream(DBOperations.READ, isRange, toBytesOrNull(key), null, store.get(key))
+  }
+
+  def getAll(keys: util.List[K]): util.Map[K, V] = {
+    val isRange = true
+    measureLatencyAndWriteToStream(DBOperations.READ, isRange, null, toBytesKey(keys.iterator()), store.getAll(keys))
+  }
+
+  def put(key: K, value: V): Unit = {
+    val isRange = true
+    measureLatencyAndWriteToStream(DBOperations.WRITE, isRange, toBytesOrNull(key), null, store.put(key, value))
+  }
+
+  def putAll(entries: util.List[Entry[K, V]]): Unit = {
+    val iter = entries.iterator
+    val isRange = true
+    measureLatencyAndWriteToStream(DBOperations.WRITE, isRange, null, toBytesKeyFromEntries(iter), store.putAll(entries))
+  }
+
+  def delete(key: K): Unit = {
+    val isRange = true
+    measureLatencyAndWriteToStream(DBOperations.DELETE, isRange, toBytesOrNull(key), null, store.delete(key))
+  }
+
+  def deleteAll(keys: util.List[K]): Unit = {
+    val isRange = true
+    measureLatencyAndWriteToStream(DBOperations.DELETE, isRange, null, toBytesKey(keys.iterator()), store.deleteAll(keys))
+  }
+
+  def range(from: K, to: K): KeyValueIterator[K, V] = {
+    val isRange = true
+    val list : util.ArrayList[K] = new util.ArrayList[K]()
+    list.add(from)
+    list.add(to)
+    measureLatencyAndWriteToStream(DBOperations.RANGE, isRange, null, toBytesKey(list.iterator()), store.range(from, to))
+  }
+
+  def all(): KeyValueIterator[K, V] = {
+    store.all()
+  }
+
+  def close(): Unit = {
+    trace("Closing.")
+
+    store.close
+  }
+
+  def flush(): Unit = {
+    trace("Flushing store.")
+
+    store.flush
+    trace("Flushed store.")
+  }
+
+
+  def toBytesKey(keys: util.Iterator[K]): util.ArrayList[Array[Byte]] = {
+    val keysInBytes = new util.ArrayList[Array[Byte]]
+    if (keys != null)
+      while(keys.hasNext()) {
+        val entry = keys.next()
+        keysInBytes.add(toBytesOrNull(entry))
+      }
+
+    keysInBytes
+  }
+
+  def toBytesKeyFromEntries(iter: util.Iterator[Entry[K, V]]): util.ArrayList[Array[Byte]] = {
+    val keysInBytes = new util.ArrayList[Array[Byte]]
+    if (iter != null)
+      while(iter.hasNext()) {
+        val entry = iter.next().getKey
+        keysInBytes.add(toBytesOrNull(entry))
+      }
+
+    keysInBytes
+  }
+
+  private def measureLatencyAndWriteToStream[R](dBOperations: Int, isRange: Boolean, singleKey: Array[Byte], keys: util.ArrayList[Array[Byte]], block: => R):R = {
+    val time1 = System.nanoTime()
+    val result = block
+    val time2 = System.nanoTime()
+    if (rng.nextInt() > samplingRatio) {
```

Comment:
Would n't it be simpler to get rid off the intermediate `return`? You can invert this logic, optionally write to the access log and return the result anyways.

Something on the the lines of:

```
long startTime =..
retVal = operation();
long endTime = ..

if (shouldLog) {
     //write to Access Log 
  }

return result
```